File: application/src/test/java/com/catalog/application/category/retrieve/get/GetCategoryByIdUseCaseTest.java
package com.catalog.application.category.retrieve.get;

import com.catalog.domain.category.Category;
import com.catalog.domain.category.CategoryGateway;
import com.catalog.domain.category.CategoryID;
import com.catalog.domain.exceptions.DomainException;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Optional;

import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class GetCategoryByIdUseCaseTest {
    @InjectMocks
    private DefaultGetCategoryByIdUseCase useCase;

    @Mock
    private CategoryGateway categoryGateway;

    @BeforeEach
    void cleanUp() {
        Mockito.reset(categoryGateway);
    }

    @Test
    public void givenAValidId_whenCallsGetCategory_shouldReturnCategory() {
        final var expectedName = "Movies";
        final var expectedDescription = "Category most watched";
        final var expectedIsActive = true;

        final var aCategory = Category.newCategory(expectedName, expectedDescription, expectedIsActive);
        final var expectedId = aCategory.getId();

        when(categoryGateway.findById(eq(expectedId)))
                .thenReturn(Optional.of(aCategory.clone()));

        final var actualCategory = useCase.execute(expectedId.getValue());

        Assertions.assertEquals(CategoryOutput.from(aCategory), actualCategory);
        Assertions.assertEquals(expectedName, actualCategory.name());
        Assertions.assertEquals(expectedDescription, actualCategory.description());
        Assertions.assertEquals(expectedIsActive, actualCategory.isActive());
        Assertions.assertEquals(expectedId, actualCategory.id());
        Assertions.assertNotNull(actualCategory.createdAt());
        Assertions.assertNotNull(actualCategory.updatedAt());
        Assertions.assertEquals(aCategory.getCreatedAt(), actualCategory.createdAt());
        Assertions.assertEquals(aCategory.getUpdatedAt(), actualCategory.updatedAt());
        Assertions.assertNull(actualCategory.deletedAt());
    }

    @Test
    public void givenAnInvalidId_whenCallsGetCategory_shouldReturnNotFound() {
        final var expectedErrorMessage = "Category with ID 123 not found.";
        final var expectedId  = CategoryID.from("123");

        when(categoryGateway.findById(eq(expectedId)))
                .thenReturn(Optional.empty());

        final var actualException = Assertions.assertThrows(
                DomainException.class,
                () -> useCase.execute(expectedId.getValue()));

        Assertions.assertEquals(expectedErrorMessage, actualException.getMessage());
    }

    @Test
    public void givenAValidId_whenGatewayThrowsException_shouldReturnException() {
        final var expectedErrorMessage = "Gateway Error";
        final var expectedId  = CategoryID.from("123");

        when(categoryGateway.findById(eq(expectedId)))
                .thenThrow(new IllegalStateException(expectedErrorMessage));

        final var actualException = Assertions.assertThrows(
                IllegalStateException.class,
                () -> useCase.execute(expectedId.getValue()));

        Assertions.assertEquals(expectedErrorMessage, actualException.getMessage());
    }
}

File: application/src/test/kotlin/com/full/admin/catalog/application/category/retrieve/get/GetCategoryByIdUseCaseTest.kt
package com.full.admin.catalog.application.category.retrieve.get

import category.Category
import category.CategoryGateway
import category.CategoryID
import exceptions.DomainException
import io.mockk.clearMocks
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(MockKExtension::class)
class GetCategoryByIdUseCaseTest {
    @InjectMockKs
    private lateinit var useCase: DefaultGetCategoryByIdUseCase

    @MockK
    private lateinit var categoryGateway: CategoryGateway

    @BeforeEach
    fun cleanUp() {
        clearMocks(categoryGateway)
    }

    @Test
    fun givenAValidId_whenCallsGetCategory_shouldReturnCategory() {
        val aCategory = Category.newCategory("Movies", "The category most watched", true)

        val expectedName = "Movies"
        val expectedDescription = "The category most watched"
        val expectedIsActive = true
        val expectedId = aCategory.id

        every { categoryGateway.findById(eq(expectedId)) } returns aCategory.clone()

        val actualCategory = useCase.execute(expectedId.value)

        Assertions.assertEquals(expectedId, actualCategory.id)
        Assertions.assertEquals(expectedName, actualCategory.name)
        Assertions.assertEquals(expectedDescription, actualCategory.description)
        Assertions.assertEquals(expectedIsActive, actualCategory.isActive)
        Assertions.assertEquals(aCategory.createdAt, actualCategory.createdAt)
        Assertions.assertEquals(aCategory.updatedAt, actualCategory.updatedAt)
        Assertions.assertEquals(aCategory.deletedAt, actualCategory.deletedAt)
        Assertions.assertEquals(CategoryOutput.from(aCategory), actualCategory)
    }

    @Test
    fun givenAInvalidId_whenCallsGetCategory_shouldReturnNotFound() {
        val expectedErrorCount = 1
        val expectedId = CategoryID.from("fake-id")
        val expectedErrorMessage = "Category with ID ${expectedId.value} was not found"

        every { categoryGateway.findById(eq(expectedId)) } returns null

        val actualException = Assertions.assertThrows(DomainException::class.java) {
            useCase.execute(expectedId.value)
        }

        Assertions.assertEquals(expectedErrorCount, actualException.getErrors().size)
        Assertions.assertEquals(expectedErrorMessage, actualException.getErrors()[0].message)
    }

    @Test
    fun givenAValidId_whenGatewayThrowsException_shouldReturnException() {
        val aCategory = Category.newCategory("Movies", "The category most watched", true)

        val expectedId = aCategory.id
        val expectedErrorMessage = "Gateway error"

        every { categoryGateway.findById(eq(expectedId)) } throws IllegalStateException(expectedErrorMessage)

        val actualException = Assertions.assertThrows(IllegalStateException::class.java) {
            useCase.execute(expectedId.value)
        }

        Assertions.assertEquals(expectedErrorMessage, actualException.message)

        verify(exactly = 1) { categoryGateway.findById(eq(expectedId)) }
    }
}